usersession/agent/rest_api_test.go

```diff
@@ -379,3 +386,288 @@ func (s *restSuite) TestServicesStopReportsTimeout(c *C) {
+		// boring stuff is checked above
+		c.Check(msg.Body[3], Equals, `Pending update of "pkg" snap`)
+		c.Check(msg.Body[4], Equals, "Close the app to avoid disruptions (7 hours left)")
+		c.Check(msg.Body[6], DeepEquals, map[string]dbus.Variant{
+			"urgency":       dbus.MakeVariant(byte(notification.NormalUrgency)),
+			"desktop-entry": dbus.MakeVariant("io.snapcraft.SessionAgent"),
+		})
+	})
+}
+
+func (s *restSuite) TestPostPendingRefreshNotificationFewMinutes(c *C) {
+	refreshInfo := &client.PendingSnapRefreshInfo{
+		InstanceName:  "pkg",
+		TimeRemaining: time.Minute * 15,
+	}
+	s.testPostPendingRefreshNotificationBody(c, refreshInfo, func(c *C, msg *dbus.Message) {
+		// boring stuff is checked above
+		c.Check(msg.Body[3], Equals, `Pending update of "pkg" snap`)
+		c.Check(msg.Body[4], Equals, "Close the app to avoid disruptions (15 minutes left)")
+		c.Check(msg.Body[6], DeepEquals, map[string]dbus.Variant{
+			"urgency":       dbus.MakeVariant(byte(notification.CriticalUrgency)),
+			"desktop-entry": dbus.MakeVariant("io.snapcraft.SessionAgent"),
+		})
+	})
+}
+
+func (s *restSuite) TestPostPendingRefreshNotificationBusyAppDesktopFile(c *C) {
+	refreshInfo := &client.PendingSnapRefreshInfo{
+		InstanceName:        "pkg",
+		BusyAppName:         "app",
+		BusyAppDesktopEntry: "pkg_app",
+	}
+	err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755)
+	c.Assert(err, IsNil)
+	desktopFilePath := filepath.Join(dirs.SnapDesktopFilesDir, "pkg_app.desktop")
+	err = ioutil.WriteFile(desktopFilePath, []byte(`
+[Desktop Entry]
+Icon=app.png
+	`), 0644)
+	c.Assert(err, IsNil)
+
+	s.testPostPendingRefreshNotificationBody(c, refreshInfo, func(c *C, msg *dbus.Message) {
+		// boring stuff is checked above
+		c.Check(msg.Body[2], Equals, "app.png")
+		c.Check(msg.Body[6], DeepEquals, map[string]dbus.Variant{
+			"urgency":       dbus.MakeVariant(byte(notification.CriticalUrgency)),
+			"desktop-entry": dbus.MakeVariant("io.snapcraft.SessionAgent"),
+		})
+	})
+}
+
+func (s *restSuite) TestPostPendingRefreshNotificationBusyAppMalformedDesktopFile(c *C) {
+	refreshInfo := &client.PendingSnapRefreshInfo{
+		InstanceName:        "pkg",
+		BusyAppName:         "app",
+		BusyAppDesktopEntry: "pkg_app",
+	}
+	err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755)
+	c.Assert(err, IsNil)
+	desktopFilePath := filepath.Join(dirs.SnapDesktopFilesDir, "pkg_app.desktop")
+	err = ioutil.WriteFile(desktopFilePath, []byte(`garbage!`), 0644)
+	c.Assert(err, IsNil)
+
+	s.testPostPendingRefreshNotificationBody(c, refreshInfo, func(c *C, msg *dbus.Message) {
+		// boring stuff is checked above
+		c.Check(msg.Body[2], Equals, "") // Icon is not provided
+		c.Check(msg.Body[6], DeepEquals, map[string]dbus.Variant{
+			"desktop-entry": dbus.MakeVariant("io.snapcraft.SessionAgent"),
+			"urgency":       dbus.MakeVariant(byte(notification.CriticalUrgency)),
+		})
+	})
+}
+
+func (s *restSuite) TestPostPendingRefreshNotificationNoNotificatinServer(c *C) {
```

Comment:
```suggestion
func (s *restSuite) TestPostPendingRefreshNotificationNoNotificationServer(c *C) {
```

Reply:
I'll send a follow-up for this.

Reply:
Sent as https://github.com/snapcore/snapd/pull/9552